text/0007-service-modules.md

```diff
@@ -0,0 +1,128 @@
+locate all the tests that need to be run when a model class changes.
+
+# Proposed Python Structure
+
+As sentry is a django application, we can leverage the
+[Django-apps](https://docs.djangoproject.com/en/4.1/ref/applications/) to act as
+a container for application services in the future. While not all services will
+need all the features of Django Applications, many will.
+
+## Django app structure
+
+We'll use 'discover' as an example for the service modules
+
+```
+src/sentry/discover
+    __init__.py
+    app.py
+    urls.py
+    models/__init__.py
+    models/discoversavedquery.py
+    endpoints/discoverquery.py
+    serializers/discoverquery.py
+    tasks/deduplicate_things.py
+
+    tests/__init__.py
+    tests/models/test_discoversavedquery.py
+    tests/endpoints/test_discoverquery.py
+    tests/serializers/test_discoverquery.py
+```
+
+In addition to the Django related modules, celery tasks, consumers and any other
+modules can be contained within a service. If a service doesn't provide
+endpoints or use models it can still benefit from the proposed structure.
+
+## Test location
+
+Tests for a service would continue to live inside the top-level `tests`
+directory. The `tests` tree would mirror continue to mirror the service + module
+structure of the application code. Sharing naming conventions should make
+running sub-sets of tests simpler to automate.
+
+## Formal entry points
+
+Service modules would use `__init__.py` to define the interface they present to
+the rest of the Sentry monolith. Having the public interface of a service
+formally defined limits the amount of entanglement the rest of the application
+can create.
+
+## Importing service internals is not allowed
+
+An important change from the present application structure is that modules
+outside of a service's scope would be *disallowed* from importing modules inside
+a service. Modules outside of a service boundary may only import the top-level
+service.
+
+Disallowing cross-service internal imports could be enforced with the
+[flake8-import-graph](https://pypi.org/project/flake8-import-graph/) extension.
+
+# Options Considered
+
+Another approach to this would be to put 'services' inside the directories of
+each 'kind'. Again using discover as an example:
+
+```
+src/sentry
+  endpoints/discover/discover_query.py
+  models/discover/discoversavedquery.py
+  tests/models/discover/test_discoversavedquery.py
+```
+
+This approach dilutes the consistency benefits, and requires a significantly
+more complex import graph rules. It also does not improve local development or
+offer benefits to CI subsetting.
+
+# Drawbacks
+
+This approach will require moving **most** of the application source code
+around. We currently store classpaths in several locations in the database. We
+may need to use data migrations to update these paths or maintain aliases for
+compatibility.
+
+Co-locating tests with application code is a potentially contentious change.
```

Comment:
I'm unsure how this line relates to the current draft. I don't see that co-located tests are recommended, above.  Unless you're calling this directory structure co-location, but this isn't the first thing people think of when they hear "co-located tests".

```
src/sentry/discover
    models/discoversavedquery.py
    tests/models/test_discoversavedquery.py
```

Test co-location usually means (something equivalent to):

```
src/sentry/discover
    models/discoversavedquery.py
    models/discoversavedquery_test.py
```

I personally like that model, but as you note it's not commonplace in python.

Reply:
I would propose to separate out the colocation discussion about tests into a separate proposal and to remove it form the scope of this. I think the test location discussion has a high bikeshedding potential but I think underpinning that largely philosophical debate are good reasons for one or the other which deserve to be discussed independently.

Either way I agree that we need a proposal for where tests go, but that we can have independently of the service boundaries.